### PR TITLE
Fix for #1059

### PR DESCRIPTION
**Pull request by trootnev**
### Database name
SQL Server

# Pull request description
Added function to get the real log backup metadata
it goes to the URL and gets it with RESTORE HEADERONLY
it is accumulated for each log backup that is supposed to be restored
later this data is iterated over during log restore making restore process more reliable.

### Describe what this PR fix
Issue #1059

### CODE REVIEW

File: internal/databases/sqlserver/log_restore_handler.go [hunk 1]
```diff
@@ -56,22 +56,43 @@ func HandleLogRestore(backupName string, untilTS string, dbnames []string, fromn
...
 		dbname := dbnames[i]
 		fromname := fromnames[i]
+		var lastPiece bool
+		var LogsMetadata []BackupProperties
+
 		for _, logBackupName := range logs {
-			ok, err := doesLogBackupContainDB(folder, logBackupName, fromname)
+			baseURL := getLogBackupURL(logBackupName, fromname)
+			basePath := getLogBackupPath(logBackupName, fromname)
+			blobs, err := listBackupBlobs(folder.GetSubFolder(basePath))
```
Comment (mialinx) [inline internal/databases/sqlserver/log_restore_handler.go:65]: What if blobs is empty list ? eg database `fromname` is not present in particular logBackupName.
Before this patch there was a `doesLogBackupContainDB` check here

File: internal/databases/sqlserver/log_restore_handler.go [hunk 2]
```diff
@@ -56,22 +56,43 @@ func HandleLogRestore(backupName string, untilTS string, dbnames []string, fromn
...
+			if err != nil {
+				return err
+			}
+			urls := buildRestoreUrls(baseURL, blobs)
+			trn, err := ListBackupProperties(db, urls, logBackupName)
```
Comment (mialinx) [inline internal/databases/sqlserver/log_restore_handler.go:70]: What `trn` stands for ?

File: internal/databases/sqlserver/log_restore_handler.go [hunk 3]
```diff
@@ -56,22 +56,43 @@ func HandleLogRestore(backupName string, untilTS string, dbnames []string, fromn
...
-			if !ok {
-				// some log backup may not contain particular database in case
-				// it was created or dropped between base backups
+			LogsMetadata = append(LogsMetadata, trn[0])
+		}
+
+		logBeginTS := time.Date(1900, 1, 1, 1, 1, 1, 1, time.UTC)
+
+		for _, bak := range LogsMetadata {
+			if bak.DatabaseName != fromname {
```
Comment (mialinx) [inline internal/databases/sqlserver/log_restore_handler.go:80]: Useless check. We collect properties only for log backups for `fromname` (line 63, 64)
There should not be properties for other databases.

File: internal/databases/sqlserver/log_restore_handler.go [hunk 4]
```diff
@@ -89,16 +110,24 @@ func restoreSingleLog(ctx context.Context,
 	logBackupName string,
 	dbname string,
 	fromname string,
-	stopAt time.Time) error {
+	stopAt time.Time,
+	lastPiece bool) error {
```
Comment (mialinx) [inline internal/databases/sqlserver/log_restore_handler.go:114]: why do we need this flag?

**Review by mialinx — CHANGES_REQUESTED** [review]

File: internal/databases/sqlserver/log_restore_handler.go (hunk 4 above)
Comment (trootnev) [reply internal/databases/sqlserver/log_restore_handler.go:114]: Not to add WITH STOPAT to every log restore operation except the last one.

**Review by mialinx — APPROVED** [review]

**Pull request merged**